compiler/optimizer/abstractinterpreter/InliningProposal.cpp

```diff
@@ -0,0 +1,275 @@
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+
+#include "optimizer/abstractinterpreter/InliningProposal.hpp"
+#include "infra/deque.hpp"
+#include "compile/Compilation.hpp"
+#include "env/Region.hpp"
+#include "infra/BitVector.hpp"
+#include "env/VerboseLog.hpp"
+#include "infra/String.hpp"
+
+TR::InliningProposal::InliningProposal(TR::Region& region, TR::IDT *idt):
+   _region(region),
+   _nodes(NULL), // Lazy Initialization of BitVector
+   _cost(0),
+   _benefit(0),
+   _idt(idt)
+   {
+   }
+
+TR::InliningProposal::InliningProposal(const InliningProposal &proposal, TR::Region& region):
+   _region(region),
+   _cost(proposal._cost),
+   _benefit(proposal._benefit),
+   _idt(proposal._idt)
+   {
+   if (proposal._nodes)
+      {
+      _nodes = new (region) TR_BitVector(proposal._nodes->getHighestBitPosition(), region);
+      *_nodes = *proposal._nodes;
+      }
+   else
+      _nodes = NULL;
+   }
+
+void TR::InliningProposal::print(TR::Compilation* comp)
+   {
+   bool traceBIProposal = comp->getOption(TR_TraceBIProposal);
+   bool verboseInlining = comp->getOptions()->getVerboseOption(TR_VerboseInlining);
+
+   if (!traceBIProposal && !verboseInlining) //no need to run the following code if neither flag is set
+      return;
+
+   TR_VerboseLog::CriticalSection vlogLock(verboseInlining);
+
+   if (!_nodes)
+      {
+      if (traceBIProposal)
+         traceMsg(comp, "Inlining Proposal is NULL\n");
+      if (verboseInlining)
+         TR_VerboseLog::writeLine(TR_Vlog_BI, "%s", "Inlining Proposal is NULL\n");
+      return;
+      }
+
+   const uint32_t numMethodsToInline =  _nodes->elementCount()-1;
+
+   TR_ASSERT_FATAL(_idt, "Must have an IDT");
+
+   TR::StringBuf header(comp->trMemory()->currentStackRegion());
+   header.appendf("#Proposal: %d methods inlined into %s, cost: %d", numMethodsToInline, _idt->getRoot()->getName(comp->trMemory()), getCost());
+
+   if (traceBIProposal)
+      traceMsg(comp, "%s\n", header.text());
+   if (verboseInlining)
+      TR_VerboseLog::writeLine(TR_Vlog_BI, "%s", header.text());
+
+   TR::deque<TR::IDTNode*, TR::Region&> idtNodeQueue(comp->trMemory()->currentStackRegion());
+   idtNodeQueue.push_back(_idt->getRoot());
+
+   //BFS
+   while (!idtNodeQueue.empty())
+      {
+      TR::IDTNode* currentNode = idtNodeQueue.front();
+      idtNodeQueue.pop_front();
+      const int32_t index = currentNode->getGlobalIndex();
+
+      if (index != -1) //do not print the root node
+         {
+         TR::StringBuf line(comp->trMemory()->currentStackRegion());
```

Comment:
Could you please reuse the `StringBuf`? This will currently create a new one for each line, but it should only be necessary to use a single instance (including for the header line above). Reusing the buffer will reduce memory allocation

Reply:
Changed it to reuse only one instance.